pgx-tests/src/tests/heap_tuple.rs

```diff
@@ -466,85 +467,94 @@ mod sql_generator_tests {
             ),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
     }
 
     #[pg_extern]
-    fn iterable_named_table() -> impl Iterator<
-        Item = (
+    fn iterable_named_table() -> TableIterator<
+        'static,
+        (
             name!(dog, ::pgx::composite_type!("Dog")),
             name!(cat, ::pgx::composite_type!("Cat")),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
     }
 
     #[pg_extern]
-    fn iterable_named_table_optional_elems() -> impl Iterator<
-        Item = (
+    fn iterable_named_table_optional_elems() -> TableIterator<
+        'static,
+        (
             name!(dog, Option<::pgx::composite_type!("Dog")>),
             name!(cat, Option<::pgx::composite_type!("Cat")>),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
     }
 
     #[pg_extern]
-    fn iterable_named_table_array_elems() -> impl Iterator<
-        Item = (
+    fn iterable_named_table_array_elems() -> TableIterator<
+        'static,
+        (
             name!(dog, Vec<::pgx::composite_type!("Dog")>),
             name!(cat, Vec<::pgx::composite_type!("Cat")>),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
     }
 
     #[pg_extern]
-    fn iterable_named_table_optional_array_elems() -> impl Iterator<
-        Item = (
+    fn iterable_named_table_optional_array_elems() -> TableIterator<
+        'static,
+        (
             name!(dog, Option<Vec<::pgx::composite_type!("Dog")>>),
             name!(cat, Option<Vec<::pgx::composite_type!("Cat")>>),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
     }
 
     #[pg_extern]
-    fn iterable_named_table_optional_array_optional_elems() -> impl Iterator<
-        Item = (
+    fn iterable_named_table_optional_array_optional_elems() -> TableIterator<
+        'static,
+        (
             name!(dog, Option<Vec<Option<::pgx::composite_type!("Dog")>>>),
             name!(cat, Option<Vec<Option<::pgx::composite_type!("Cat")>>>),
         ),
     > {
-        Vec::new().into_iter()
+        TableIterator::new(Vec::new().into_iter())
```

Comment:
This makes me wonder if the winning move is I: IntoIterator?

Reply:
I recall this created some curious errors in some existing code. 🤔 It is something I initially tried.